### PR TITLE
Bn/cnx#1461/update sync pipeline

### PR DESCRIPTION
**Pull request by c0ldcalamari**
- add custom resource to repo used to trigger pipeline
- let pipeline use custom resource as trigger
- instructions on how to stand up pipeline
- tests for resource

### CODE REVIEW

File: pipeline-sync.yml [hunk 1]
```diff
@@ -1,60 +1,77 @@
...
+  - name: ((book-repo))
     type: git
     source:
       branch: main
-      uri: ((osbook-git-uri))
-      private_key: ((ce-github-private-key))
+      <<: *repo-authorization
 
-  - name: synced-osbook
+  - name: synced-((book-repo))
```
Comment (tomjw64) [inline pipeline-sync.yml:31]: Did it turn out that this resource is necessary? Is `put: ((book-repo))` possible?

File: resource_trigger/check.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    # Get versions from uuid via archive api
+    versions = []
+    for legacy_id in legacy_ids:
+        try:
+            uuid = BOOK_UUIDS[legacy_id]
+        except KeyError as error:
+            msg(error, file=sys.stderr)
+            sys.exit(1)
+        link = f"https://{archive_server}/contents/{uuid}"
```
Comment (tomjw64) [inline resource_trigger/check.py:52]: I'm pretty sure that this provides the `latest successfully baked` version and not the `latest` version of the collection. Have you seen something that indicates otherwise? For similar reasons, we do not use the `version` field of this API endpoint with neb to determine the latest version to pull for CMs. This may need to use the `extras` endpoint.

File: pipeline-sync.yml (hunk 1 above)
Comment (c0ldcalamari) [reply pipeline-sync.yml:31]: yes, when trying to read and write to the same git resource - i found that the best way was to use two resources. 
if we only use `osbook` and error will occur (I think it could have been handled by running git commands, but this was much cleaner)

Comment (c0ldcalamari) [reply pipeline-sync.yml:31]: (the naming of the `put` resource doesn't seem right but..) 
I believe this was the error:
Screen Shot 2021-03-09 at 1 48 11 PM

File: resource_trigger/check.py (hunk 2 above)
Comment (c0ldcalamari) [reply resource_trigger/check.py:52]: ah interesting.. so when a 'publish' is done by neb, it doesn't mean it gets baked..?

Comment (c0ldcalamari) [reply resource_trigger/check.py:52]: diff between - 
'latest successfully baked'
'latestVersion'
'headVersion'

example link: https://archive.cnx.org/extras/9d8df601-4f12-4ac1-8224-b450bf739e5f@6.10

File: pipeline-sync.yml (hunk 1 above)
Comment (tomjw64) [reply pipeline-sync.yml:31]: isn't this solved by using `force`, regardless of the resource used to `put`?

Comment (c0ldcalamari) [reply pipeline-sync.yml:31]: I think I mis-answered your first question - 
"Did it turn out that this resource is necessary? Is put: ((book-repo)) possible?"
This was necessary with the detail that we were working with more than one branch, and that we would always pull from main branch but not `put` (push) back to the main branch. 

Initially is was unclear having thrown a 'staging/qa' branch in earlier discussions.  I made more assumptions dealing with a `put` back to the a branch that was not `main`. 
 I believe that error did because I wasn't using the `force` param, so that definitely makes sense. 

but (with confirmation/ clarification from check in) in the end, that resource isn't necessary. (I'm removing it!) 
Since we will only be pull and pushing to the main branch. - 

tldr; will be updated to use one git (book) resource

File: resource_trigger/test_sync_trigger.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    mocker.patch(
+        'check.get_sync_file',
+        return_value=sync_file
+    )
+
+    instream = fake_input_stream()
+    versions = check._check(instream)[0]['versions']
+    number_of_actual_versions = len(versions.split(','))
+    number_of_expected_versions = 4
+    assert number_of_expected_versions == number_of_actual_versions
```
Comment (tomjw64) [inline resource_trigger/test_sync_trigger.py:46]: Can you mock out the requests made in the tests, so that you can assert the exact output?

File: resource_trigger/test_sync_trigger.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    def make_stream(json_obj):
+        stream = io.StringIO()
+        json.dump(json_obj, stream)
+        stream.seek(0)
+        return stream
+
+    def make_input():
+        return {
+            "source": {
+                "archive_server": "qa.cnx.org",
```
Comment (tomjw64) [inline resource_trigger/test_sync_trigger.py:18]: Might be nice to have a unit test for the archive server resolution

**Review by tomjw64 — COMMENTED** [review]
Code itself looks great. Comments on some small things.

File: resource_trigger/check.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    # headers = {'Authorization': 'token ' + token}
+    # endpoint = f"https://raw.githubusercontent.com/openstax/{repo}/main/archive-syncfile"
+    # resp = requests.get(endpoint, headers=headers)
+
+    # if resp.status_code != 200:
+    #     msg(f"Error: Unable to get archive-syncfile for {repo}")
+    #     sys.exit(1)
+
+    # sync_file = resp.content.decode()
```
[the remark this answers is not in the document]
Comment (c0ldcalamari) [reply resource_trigger/check.py]: removed!

File: resource_trigger/test_sync_trigger.py (hunk 4 above)
Comment (c0ldcalamari) [reply resource_trigger/test_sync_trigger.py:18]: What do you mean by 'archive server resolution"? do you mean a test for `determine archive server` function?

Comment (tomjw64) [reply resource_trigger/test_sync_trigger.py:20]: yup!

**Review by tomjw64 — COMMENTED** [review]
This may still be getting work on it, but I'll re-review since its in the review column, just in case. Looks like we may want to add `__pycache__` files to `.gitignore`. Additionally, it looks like `test_check_type` and `test_version_format` still make unmocked requests to an archive `/extras` endpoint.

**Comment by c0ldcalamari** [comment]
@tomjw64 ah, I had added two separate tests to address the two requests made - `test_github_request` and `test_archive_request` but I can incorporate the mocks into the other tests.

will add .__pycache__ to `.gitignore`

**Comment by tomjw64** [comment]
Yeah, it's less about testing with the mock data, and more about not having unit tests fail depending on whether or not archive servers/github are experiencing issues.

**Comment by c0ldcalamari** [comment]
@tomjw64 that makes sense, all done!

**Review by tomjw64 — APPROVED** [review]

**Comment by c0ldcalamari** [comment]
Requirements for Testing:

- Set up pipeline from Content Synchronization - ref: [Quick Start](https://github.com/openstax/content-synchronizer#quick-start) - for [osbooks-college-algebra-bundle](https://github.com/openstax/osbooks-college-algebra-bundle) using `archive-qa.cnx.org`

```
fly -t local-concourse set-pipeline \
--config pipeline-sync.yml \
--pipeline sync-pipeline \
--load-vars-from vars.yml \
--var book-repo=osbooks-college-algebra-bundle \
--var archive-server=qa
```
Note: The change for the pipeline is using the [resource trigger](https://github.com/openstax/content-synchronizer/blob/861e6596cf6397ecdaa79ff0b827840cc2399b66/pipeline-sync.yml#L11) as opposed to using the time resource. 

- Latest version of one of the books in the [archive-syncfile](https://github.com/openstax/osbooks-college-algebra-bundle/blob/main/archive-syncfile), pulled by neb from QA. 
```
neb get -r qa col11759 latest
```

## Given 
The custom resource ([resource trigger](https://github.com/openstax/content-synchronizer/tree/main/resource_trigger)) for the sync pipeline
## When
A new head version is detected (on archive after a neb publish)
## Then
The resource (resource_trigger) will trigger the pipeline and sync the content, changes can be observed on the `main` branch of [osbooks-college-algebra-bundle](https://github.com/openstax/osbooks-college-algebra-bundle)
